ast.ts

```diff
@@ -39,7 +39,18 @@ export type Expr<A> =
   | {  a?: A, tag: "lookup", obj: Expr<A>, field: string }
   | {  a?: A, tag: "index", obj: Expr<A>, index: Expr<A> }
   | {  a?: A, tag: "method-call", obj: Expr<A>, method: string, arguments: Array<Expr<A>> }
-  | {  a?: A, tag: "construct", name: string }
+  | {  a?: A, tag: "construct", name: string, arguments: Array<Expr<A>> }
+  // ast for list-comp
+  | {  a?: A, tag: "list-comp", left: Expr<A>, elem: Expr<A>, iterable: Expr<A>, cond?: Expr<A>, iterable_cond?: Expr<A>, body?: Array<Stmt<A>> }
+
+  // Possible Typechecks here ::
+  // 1. iterable.a.tag = list
+  // 2. cond is optional and cond.a.tag = bool and cond.tag = "if"
+  // 3. elem.tag = id
```

Comment:
Why have elem be an Expr? Why not a string? Or a LHS?

Reply:
These comments are actually part of our design decisions from Week 1 of the project. Although the comments are now removed, our though process was as follows: [x for a in range(5)], where x is represented by elem. Now this x can be anything ranging from a variable or a string or some mathematical value (involving unary or binop) or a function call. And every one of these cases comes under expression.